### PR TITLE
AEA-740 dialogues support for packages/connections

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes
add dialogues support for connections: http_client/server, local, gym

Also:
- fix gym examples for new dialogues etc
- update for incoming message should check both sender and counterparty
- fixes remaining tests using upgraded connections
- fixes gym and http skills and examples
- stricter rules on message processing in dialogues update

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: examples/gym_ex/proxy/env.py [hunk 1]
```diff
@@ -115,7 +119,10 @@ def reset(self) -> None:
         """
         if not self._agent.multiplexer.is_connected:
             self._connect()
-        gym_msg = GymMessage(performative=GymMessage.Performative.RESET)
+        gym_msg = GymMessage(
+            dialogue_reference=self._dialogues.new_self_initiated_dialogue_reference(),
+            performative=GymMessage.Performative.RESET,
+        )
```
Comment (DavidMinarsch) [inline examples/gym_ex/proxy/env.py:125]: Gym dialogue structure is as follows:

```
reply:
  reset: [status]
  status: [act, close, reset]
  act: [percept]
  percept: [act, close, reset]
  close: []
```
you need to ensure this is followed here.

Comment (DavidMinarsch) [reply examples/gym_ex/proxy/env.py:125]: So after sending `reset` we expect the next message to be a `status`; etc.

File: packages/fetchai/connections/gym/connection.py [hunk 2]
```diff
@@ -58,6 +61,55 @@ def __init__(self, address: Address, gym_env: gym.Env):
...
+        :return: new response message with all dialogue details.
+        """
+        response_message.set("target", incoming_message.message_id)
+        response_message.set("message_id", incoming_message.message_id + 1)
+        response_message.counterparty = incoming_message.counterparty
+        dialogue.update(response_message)
+        response_message = copy.deepcopy(
+            response_message
+        )  # TODO: fix; need to copy atm to avoid overwriting "is_incoming"
+        return response_message
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/gym/connection.py:112]: `target, message_id and dialogue_reference` should be set on message construction (line 114 onwards), not here

File: packages/fetchai/connections/http_client/connection.py [hunk 3]
```diff
@@ -263,16 +326,16 @@ def to_envelope(
         """
         context = EnvelopeContext(connection_id=connection_id)
         http_message = HttpMessage(
-            dialogue_reference=http_request_message.dialogue_reference,
-            target=http_request_message.target,
-            message_id=http_request_message.message_id,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http_client/connection.py:268]: No; this should stay! The constructor is the right place to set these. The "update" approach should only be used for the newly added functionality (is_incoming, copy message, counterparty).

File: packages/fetchai/connections/http_client/connection.py [hunk 4]
```diff
@@ -102,7 +107,55 @@ async def connect(self, loop: AbstractEventLoop) -> None:
...
+        Create a response message  with all dialog ue details.
+
+        :param response_message: message to be sent
+        :param incoming_message: message that response constructed for
+        :param dialogue: a dialog for messages
+
+        :return: new response message with all dialogue details.
+        """
+        response_message.set("target", incoming_message.message_id)
+        response_message.set("message_id", incoming_message.message_id + 1)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http_client/connection.py:150]: set in constructor, not here

File: packages/fetchai/connections/http_client/connection.py [hunk 5]
```diff
@@ -102,7 +107,55 @@ async def connect(self, loop: AbstractEventLoop) -> None:
...
+
+    @staticmethod
+    def _set_response_dialogue_references(
+        response_message: HttpMessage,
+        incoming_message: HttpMessage,
+        dialogue: HttpDialogue,
+    ) -> HttpMessage:
+        response_message.set(
+            "dialogue_reference", dialogue.dialogue_label.dialogue_reference
+        )
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http_client/connection.py:139]: see below

Comment (solarw) [reply packages/fetchai/connections/http_client/connection.py]: done

File: packages/fetchai/connections/http_client/connection.py (hunk 4 above)
Comment (solarw) [reply packages/fetchai/connections/http_client/connection.py]: done

File: packages/fetchai/connections/http_client/connection.py (hunk 3 above)
Comment (solarw) [reply packages/fetchai/connections/http_client/connection.py:268]: done

File: packages/fetchai/connections/gym/connection.py (hunk 2 above)
Comment (solarw) [reply packages/fetchai/connections/gym/connection.py]: done

File: examples/gym_ex/proxy/env.py (hunk 1 above)
Comment (solarw) [reply examples/gym_ex/proxy/env.py:145]: need to discuss

**Pull request merged**